flasc/energy_ratio/energy_ratio.py

```diff
@@ -115,30 +119,38 @@ def _compute_energy_ratio_single(df_,
     # Group df_
     df_ = (df_
         .filter(pl.all_horizontal(pl.col(bin_cols_with_df_name).is_not_null())) # Select for all bin cols present
-        .groupby(bin_cols_with_df_name, maintain_order=True)
+        .group_by(bin_cols_with_df_name, maintain_order=True)
         .agg([pl.mean("pow_ref"), pl.mean("pow_test"),pl.count()])
-        )
-    
+
+        # Enforce that each ws/wd bin combination has to appear in all dataframes
+        .filter(pl.count().over(bin_cols_without_df_name) == num_df)
+
+    )
     # Determine the weighting of the ws/wd bins
 
     if df_freq_pl is None:
         # Determine the weights per bin as either the min or sum count
-        df_ = (df_
-            .with_columns(
-                [
-                    # Get the weighting by counts
-                    pl.col('count').min().over(bin_cols_without_df_name).alias('weight') if weight_by == 'min' else
-                    pl.col('count').sum().over(bin_cols_without_df_name).alias('weight')
-                ]
-            )
+        df_freq_pl = (df_
+            .select(bin_cols_without_df_name+['count'])
+            .group_by(bin_cols_without_df_name)
+            .agg([pl.min('count') if weight_by == 'min' else pl.sum('count')])
+            .rename({'count':'weight'})
         )
     
-    else:
-        # Use the weights in df_freq_pl directly
-        df_ = (df_.join(df_freq_pl, on=['wd_bin','ws_bin'], how='left')
-              .with_columns(pl.col('weight').fill_null(strategy="zero"))
-        )
+    df_ = (df_.join(df_freq_pl, on=['wd_bin','ws_bin'], how='left')
+            .with_columns(pl.col('weight'))# .fill_null(strategy="zero"))
+    )
+
+    # Check if all the values in the weight column are null
+    if df_['weight'].is_null().all():
+        raise RuntimeError("None of the ws/wd bins in data appear in df_freq")
+    
+    # Check if any of the values in the weight column are null
+    if df_['weight'].is_null().any():
+        warnings.warn('Some bins in data are not in df_freq and will get 0 weight')
 
+    # Fill the null values with zeros
+    df_= df_.with_columns(pl.col('weight').fill_null(strategy="zero"))
 
```

Comment:
I don't think this affects any of the computed energy ratio values, so it likely isn't necessary, but what do you think about normalizing 'weight' so it sums to 1? The only place where this might have an impact is when overlaying weights on the energy ratio plots.

Reply:
Maybe like this?